system/jlib/jfile.cpp

```diff
@@ -7643,3 +7643,110 @@ IAPICopyClient * createApiCopyClient(IStorageApiInfo * source, IStorageApiInfo *
     }
     return nullptr;
 }
+
+
+class CBlockedFileIO;
+static thread_local void *blockedFileIOBuffer = nullptr;
+static thread_local offset_t blockedFileIOLastReadPos = (offset_t)-1;
+static thread_local CBlockedFileIO *blockedFileIOLastReadCaller = nullptr;
+class CBlockedFileIO : public CSimpleInterfaceOf<IFileIO>
+{
+    Owned<IFileIO> io;
+    offset_t fileSize = 0;
+    size32_t blockSize = 0;
+public:
+    CBlockedFileIO(IFileIO *_io, size32_t _blockSize) : io(_io), blockSize(_blockSize)
+    {
+        fileSize = io->size();
+    }
+    ~CBlockedFileIO()
+    {
+        blockedFileIOLastReadCaller = nullptr;
+    }
+    static bool onThreadTerm(bool isPooled)
+    {
+        free(blockedFileIOBuffer);
+        blockedFileIOBuffer = nullptr;
+        return false;
+    }
+    virtual size32_t read(offset_t pos, size32_t len, void *data) override
+    {
+        if (len > blockSize || pos+len > fileSize || len==0)
+            return io->read(pos, len, data);
+        if (!blockedFileIOBuffer)
+        {
+            blockedFileIOBuffer = malloc(blockSize);
```

Comment:
This has problems if there are multiple CBlockedFileIO objects with different blockSizes.  Currently it would potentially corrupt memory.  If a thread local size was used instead you could prevent the corruption, but would the block size would be determined by whichever file was first read on that thread.
I think the idea may be fundamentally flawed for multithreaded access.